File: Arrays/53-Maximum-Subarray/JavaScript/53-Maximum-Subarray-Optimal.js
/**
 * Problem: 53. Maximum Subarray
 * Approach: Optimal - Kadane's Algorithm
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */

var maxSubArray = function(nums) {
    let current = nums[0];
    let maxi = nums[0];

    for (let i = 1; i < nums.length; i++) {
        current = Math.max(current + nums[i], nums[i]);
        maxi = Math.max(maxi, current);
    }

    return maxi;
};

File: Arrays/53-Maximum-Subarray/Python/53-Maximum-Subarray-Optimal.py
"""
Problem: 53. Maximum Subarray
Approach: Optimal - Kadane's Algorithm

Time Complexity: O(n)
Space Complexity: O(1)
"""


class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        current = nums[0]
        maxi = nums[0]

        for i in range(1, len(nums)):
            current = max(current + nums[i], nums[i])
            maxi = max(maxi, current)

        return maxi